### PR TITLE
Merge WebKit swift overlay into WebKit.framework

### PR DESCRIPTION
**Pull request by emw-apple**
#### 36392f6620ac514db4d182a707b719ef9aa9768f
<pre>
Merge WebKit swift overlay into WebKit.framework
<a href="https://bugs.webkit.org/show_bug.cgi?id=280734">https://bugs.webkit.org/show_bug.cgi?id=280734</a>
<a href="https://rdar.apple.com/119206029">rdar://119206029</a>

Reviewed by Alexey Proskuryakov, Andy Estes, and Richard Robinson.

Instead of building a separate libswiftWebKit.dylib containing Swift
wrappers of WebKit APIs, implement those wrappers in WebKit&apos;s main
framework target. This frees up WebKit&apos;s Cocoa port to implement new
APIs in Swift which call into ObjC++ internals.

First, move the target membership of the overlay Swift sources to the
main target. Copy over relevant build settings from the overlay target
and from PAL, where we already have support for writing Swift wrappers.

The effect is that the main WebKit target now installs a
`WebKit.swiftmodule` into the OS, which is loaded when the Swift
compiler sees &quot;import WebKit&quot; and contains metadata telling the compiler
to link against WebKit.framework.

Next, retain binary compatibility for existing Swift clients of WebKit,
who have a runtime dependency on libswiftWebKit today. This works by:

- Emitting magic $ld$previous symbols into WebKit&apos;s const table for each
  symbol we are migrating from libswiftWebKit. Dyld uses the metadata in
  these symbols to fix up libswiftWebKit bindings to WebKit bindings.

- Installing a libswiftWebKit.dylib symlink
  which points to WebKit. Binaries which load libswiftWebKit will load
  WebKit directly. The dylib is moved to the system cryptex, so that it
  is picked up when building the dyld the shared cache. In the OS, the
  symlink will be resident in the shared cache at
  /usr/local/lib/libswiftWebKit.dylib (its old install name), and
  entirely removed from filesystem. The symlink is not installed in
  downlevels, which have never built the overlay.

Finally, now that WebKit produces its own swiftmodule, change the
WebKitSwiftOverlay project to produce a root with an empty
/usr/lib/swift directory. This avoids overlapping content in the OS
build, and means that any new Swift programs that use WebKit will link
directly to the main framework.

Eventually, we can stop building WebKitSwiftOverlay and delete it. There
are some tests in this project which aren&apos;t run by anyone currently, but
may be valuable to move to API tests once we have a viable way to write
API tests in Swift.

There are three follow-up changes needed to improve build coverage of
the overlay APIs:

- Declaring needed Network.framework SPI so that we can build the
  portion of the overlay that uses Network interfaces in public builds.
  (<a href="https://bugs.webkit.org/show_bug.cgi?id=280911)">https://bugs.webkit.org/show_bug.cgi?id=280911)</a>

- Fixing availability issues in tvOS and watchOS builds, to support
  building the overlay on those platforms. (<a href="https://bugs.webkit.org/show_bug.cgi?id=280912)">https://bugs.webkit.org/show_bug.cgi?id=280912)</a>
  It&apos;s not currently built there at all, but it will be nice to have
  parity on all platforms to support future Swift adoption.

- Fixing actor-isolation warnings in concurrent code. Currently, these
  warnings are ignored by the Swift overlay project in production
  builds. (<a href="https://bugs.webkit.org/show_bug.cgi?id=280734)">https://bugs.webkit.org/show_bug.cgi?id=280734)</a>

* Configurations/CommonBase.xcconfig: Satisfy style checker by making
  OTHER_SWIFT_FLAGS inherit, indicating that it&apos;s a list-style variable
  that is supposed to be appended to by other xcconfigs.
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKit/Configurations/WebKit.xcconfig: In addition to migrated
  build settings, set SWIFT_LIBRARY_LEVEL because XCBuild can&apos;t infer
  whether WebKit is public or private due to its cryptex install path.
  Also set ASSETCATALOG_COMPILER_GENERATE_ASSET_SYMBOLS to avoid some
  generated Swift code as a side effect.
* Source/WebKit/Configurations/WebKitSwift.xcconfig:
* Source/WebKit/Shared/API/Cocoa/WebKit.m:
* Source/WebKit/SwiftOverlay/SwiftOverlay/install-swiftmodules.sh:
* Source/WebKit/SwiftOverlay/WebKitSwiftOverlay.xcodeproj/project.pbxproj:
* Source/WebKit/UIProcess/API/Cocoa/ObjectiveCBlockConversions.swift: Renamed from Source/WebKit/SwiftOverlay/SwiftOverlay/ObjectiveCBlockConversions.swift.
(ObjCBlockConversion.boxingNilAsAnyForCompatibility(_:Error:Error:)):
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftOverlayMacros.h: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/285750@main">https://commits.webkit.org/285750@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/2e4282845845607f43b932acc797da922eb2aeb0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73601 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53030 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26412 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77910 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24839 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62163 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/815 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57872 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16269 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76668 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48023 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63335 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38278 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44612 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23172 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66351 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21169 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79489 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/918 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/397 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66234 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1060 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63344 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65515 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9375 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7556 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11362 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/882 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3632 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/911 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/898 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/917 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/SwiftOverlay/SwiftOverlay/install-swiftmodules.sh [hunk 1]
```diff
@@ -29,16 +29,6 @@ set -e
...
-if [ ! -d "$srcmodule" ]; then
-    echo "Cannot find Swift module at $srcmodule" >&2
-    exit 1
-fi
-
-mkdir -p "$dstpath"
-cp -r "$srcmodule" "$dstpath"
+# This project used to install a libswiftWebKit.dylib to this directory. Until it stops building
+# altogether, produce an empty directory to satisfy the OS build system.
+mkdir -p "${INSTALL_ROOT}/usr/lib/swift"
```
Comment (aproskuryakov) [inline Source/WebKit/SwiftOverlay/SwiftOverlay/install-swiftmodules.sh:34]: This seems only needed when building for OS, not in local builds and not in downlevels or STP.

Comment (emw-apple) [reply Source/WebKit/SwiftOverlay/SwiftOverlay/install-swiftmodules.sh:34]: It's not documented publicly, but the WebKitSwiftOverlay project is only built and installed for OS. So unlike the script phase in WebKit.xcodeproj, I don't think we _need_ a check here, though we could add one for sanity.

**Review by aestes — APPROVED** [review]

**Review by rr-codes — APPROVED** [review]

File: Source/WebKit/Configurations/BaseTarget.xcconfig [hunk 2]
```diff
@@ -26,7 +26,9 @@
 
 FRAMEWORK_SEARCH_PATHS = "$(SDK_DIR)$(UMBRELLA_FRAMEWORKS_DIR)";
 
-SYSTEM_FRAMEWORK_SEARCH_PATHS = $(inherited) $(WK_PRIVATE_SDK_DIR)$(SYSTEM_LIBRARY_DIR)/PrivateFrameworks $(SDK_DIR)$(SYSTEM_LIBRARY_DIR)/Frameworks;
+// On internal SDKs, try /System/Library/Frameworks first to work around rdar://134956870.
```
Comment (aproskuryakov) [inline Source/WebKit/Configurations/BaseTarget.xcconfig:29]: Is this a problem for downlevels, where we very much don't want to load e.g. WebCore from /S/L/F?

File: Source/WebKit/UIProcess/API/Cocoa/WebKitSwiftOverlay.swift [hunk 3]
```diff
@@ -23,8 +23,11 @@
  * THE POSSIBILITY OF SUCH DAMAGE.
  */
 
-@_exported import WebKit
+#if !os(tvOS) && !os(watchOS)
+
+#if USE_APPLE_INTERNAL_SDK
```
Comment (aproskuryakov) [inline Source/WebKit/UIProcess/API/Cocoa/WebKitSwiftOverlay.swift:28]: Just to confirm, is it OK to use WTF macros here?

Comment (emw-apple) [reply Source/WebKit/UIProcess/API/Cocoa/WebKitSwiftOverlay.swift:28]: WTF macros don't work, but this is defined in CommonBase.xcconfig: https://github.com/WebKit/WebKit/blob/f55d6cc96b847b1a4c5b1ed4d5405d14dd601126/Configurations/CommonBase.xcconfig#L52-L54

File: Source/WebKit/Configurations/BaseTarget.xcconfig (hunk 2 above)
Comment (emw-apple) [reply Source/WebKit/Configurations/BaseTarget.xcconfig:29]: Good question. Downlevels have the StagedFrameworks/ path given in FRAMEWORK_SEARCH_PATHS, which is declared above on L27. (`$UMBRELLA_FRAMEWORKS_DIR` is the overridden frameworks dir on downlevels). So they end up with a compiler invocation like:

```
clang ... \
    -F<SDKROOT>/System/Cryptexes/App/System/Library/StagedFrameworks/Safari
    -iframework '<SDKROOT>/System/Library/Frameworks' \
    -iframework '<SDKROOT>/System/Library/PrivateFrameworks' \
```

and `-F` frameworks take precedence over `-iframework` frameworks.

**Comment by emw-apple** [comment]
@aproskuryakov asked me about this in person, so for posterity, here's what one of the generated .swiftinterface files looks like.

arm64-apple-macos.swiftinterface:
```swift
// swift-interface-format-version: 1.0
// swift-compiler-version: Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
// swift-module-flags: -target arm64-apple-macos14.5 -enable-objc-interop -enable-library-evolution -swift-version 5 -enforce-exclusivity=checked -O -library-level api -enable-experimental-concurrency -enable-bare-slash-regex -module-name WebKit
import Swift
@_exported import WebKit
import _Concurrency
import _StringProcessing
import _SwiftConcurrencyShims
@available(iOS 14.0, macOS 11.0, *)
extension WebKit.WKPDFConfiguration {
  @_Concurrency.MainActor(unsafe) public var rect: CoreFoundation.CGRect? {
    get
    set
  }
}
@available(iOS 14.0, macOS 11.0, *)
extension WebKit.WKWebView {
  @_Concurrency.MainActor(unsafe) public func callAsyncJavaScript(_ functionBody: Swift.String, arguments: [Swift.String : Any] = [:], in frame: WebKit.WKFrameInfo? = nil, in contentWorld: WebKit.WKContentWorld, completionHandler: ((Swift.Result<Any, any Swift.Error>) -> Swift.Void)? = nil)
  @_Concurrency.MainActor(unsafe) public func createPDF(configuration: WebKit.WKPDFConfiguration = .init(), completionHandler: @escaping (Swift.Result<Foundation.Data, any Swift.Error>) -> Swift.Void)
  @_Concurrency.MainActor(unsafe) public func createWebArchiveData(completionHandler: @escaping (Swift.Result<Foundation.Data, any Swift.Error>) -> Swift.Void)
  @_Concurrency.MainActor(unsafe) public func evaluateJavaScript(_ javaScript: Swift.String, in frame: WebKit.WKFrameInfo? = nil, in contentWorld: WebKit.WKContentWorld, completionHandler: ((Swift.Result<Any, any Swift.Error>) -> Swift.Void)? = nil)
  @_Concurrency.MainActor(unsafe) public func find(_ string: Swift.String, configuration: WebKit.WKFindConfiguration = .init(), completionHandler: @escaping (WebKit.WKFindResult) -> Swift.Void)
}
@available(iOS 15.0, macOS 12.0, *)
extension WebKit.WKWebView {
  #if compiler(>=5.3) && $AsyncAwait
  @_Concurrency.MainActor(unsafe) public func callAsyncJavaScript(_ functionBody: Swift.String, arguments: [Swift.String : Any] = [:], in frame: WebKit.WKFrameInfo? = nil, contentWorld: WebKit.WKContentWorld) async throws -> Any?
  #endif
  #if compiler(>=5.3) && $AsyncAwait
  @_Concurrency.MainActor(unsafe) public func pdf(configuration: WebKit.WKPDFConfiguration = .init()) async throws -> Foundation.Data
  #endif
  #if compiler(>=5.3) && $AsyncAwait
  @_Concurrency.MainActor(unsafe) public func evaluateJavaScript(_ javaScript: Swift.String, in frame: WebKit.WKFrameInfo? = nil, contentWorld: WebKit.WKContentWorld) async throws -> Any?
  #endif
  #if compiler(>=5.3) && $AsyncAwait
  @_Concurrency.MainActor(unsafe) public func find(_ string: Swift.String, configuration: WebKit.WKFindConfiguration = .init()) async throws -> WebKit.WKFindResult
  #endif
}
@available(iOS 18.4, macOS 15.4, visionOS 2.4, *)
extension WebKit.WKWebExtensionController {
  @_Concurrency.MainActor(unsafe) public func didCloseTab(_ closedTab: any WebKit.WKWebExtensionTab, windowIsClosing: Swift.Bool = false)
  @_Concurrency.MainActor(unsafe) public func didActivateTab(_ activatedTab: any WebKit.WKWebExtensionTab, previousActiveTab previousTab: (any WebKit.WKWebExtensionTab)? = nil)
  @_Concurrency.MainActor(unsafe) public func didMoveTab(_ movedTab: any WebKit.WKWebExtensionTab, from index: Swift.Int, in oldWindow: (any WebKit.WKWebExtensionWindow)? = nil)
}
@available(iOS 18.4, macOS 15.4, visionOS 2.4, *)
extension WebKit.WKWebExtensionContext {
  @_Concurrency.MainActor(unsafe) public func didCloseTab(_ closedTab: any WebKit.WKWebExtensionTab, windowIsClosing: Swift.Bool = false)
  @_Concurrency.MainActor(unsafe) public func didActivateTab(_ activatedTab: any WebKit.WKWebExtensionTab, previousActiveTab previousTab: (any WebKit.WKWebExtensionTab)? = nil)
  @_Concurrency.MainActor(unsafe) public func didMoveTab(_ movedTab: any WebKit.WKWebExtensionTab, from index: Swift.Int, in oldWindow: (any WebKit.WKWebExtensionWindow)? = nil)
}
```

The supported deployment target is represented by the `-target arm64-apple-macos14.5` module flag.  Availability versions are language attributes, so they are present, but conditional compilation definitions are not—they're resolved at build time. There's one .swiftinterface file per target triple, so `#if os(...)` conditions are also factored out.

**Review by aproskuryakov — APPROVED** [review]

**Comment by emw-apple** [comment]
The latest revision moves the bincompat symlink creation step to `WebKitSwiftOverlay.xcodeproj`. This is to satisfy Apple's internal OS build system, which expects that this project continues to produce `./usr/lib/swift/libswiftWebKit.dylib` and subpaths of `./usr/lib/swift/WebKit.swiftoverlay/`.

We can still eventually remove the Swift overlay project entirely, but it will need to wait until after the OS build has integrated this change.

**Comment by emw-apple** [comment]
> The latest revision moves the bincompat symlink creation step to `WebKitSwiftOverlay.xcodeproj`. This is to satisfy Apple's internal OS build system, which expects that this project continues to produce `./usr/lib/swift/libswiftWebKit.dylib` and subpaths of `./usr/lib/swift/WebKit.swiftoverlay/`.
> 
> We can still eventually remove the Swift overlay project entirely, but it will need to wait until after the OS build has integrated this change.

JK, it looks like we will not need this after all. Reverted back to the previous patch.

**Pull request merged**